core/src/main/java/jenkins/security/ApiTokenProperty.java

```diff
@@ -176,29 +317,153 @@ public String getDisplayName() {
+                // recommended way
+                return null;
+            }
+
+            return forceNewInstance(user, true);
         }
+        
+        private ApiTokenProperty forceNewInstance(User user, boolean withLegacyToken) {
+            if(withLegacyToken){
+                return new ApiTokenProperty(API_KEY_SEED.mac(user.getId()));
+            }else{
+                return new ApiTokenProperty(null);
+            }
+        }
+    
+        // for Jelly view
+        @Restricted(NoExternalUse.class)
+        public boolean hasCurrentUserRightToGenerateNewToken(User propertyOwner){
+            if (ADMIN_CAN_GENERATE_NEW_TOKENS && Jenkins.get().hasPermission(Jenkins.ADMINISTER)) {
+                return true;
+            }
+
+            User currentUser = User.current();
+            if (currentUser == null) {
+                // Anonymous
+                return false;
+            }
+
+            if (Jenkins.getAuthentication() == ACL.SYSTEM) {
+                // SYSTEM user is always eligible to see tokens
+                return true;
+            }
 
+            return User.idStrategy().equals(propertyOwner.getId(), currentUser.getId());
+        }
+
+        /**
+         * @deprecated use {@link #doGenerateNewToken(User, StaplerResponse, String)} instead
+         */
+        @Deprecated
         @RequirePOST
         public HttpResponse doChangeToken(@AncestorInPath User u, StaplerResponse rsp) throws IOException {
+            LOGGER.log(Level.WARNING, "Deprecated action /changeToken used, consider using /generateNewToken instead");
+
+            u.checkPermission(Jenkins.ADMINISTER);
+
             ApiTokenProperty p = u.getProperty(ApiTokenProperty.class);
-            if (p==null) {
-                p = newInstance(u);
+            if (p == null) {
+                p = forceNewInstance(u, true);
                 u.addProperty(p);
             } else {
+                // even if the user does not have legacy token, this method let some legacy system to regenerate one
                 p.changeApiToken();
             }
+            
             rsp.setHeader("script","document.getElementById('apiToken').value='"+p.getApiToken()+"'");
-            return HttpResponses.html(p.hasPermissionToSeeToken() 
-                    ? Messages.ApiTokenProperty_ChangeToken_Success() 
+            return HttpResponses.html(p.hasPermissionToSeeToken()
+                    ? Messages.ApiTokenProperty_ChangeToken_Success()
                     : Messages.ApiTokenProperty_ChangeToken_SuccessHidden());
         }
-    }
 
+        @RequirePOST
+        public HttpResponse doGenerateNewToken(@AncestorInPath User u, StaplerResponse rsp, @QueryParameter String newTokenName) throws IOException {
+            if(!hasCurrentUserRightToGenerateNewToken(u)){
+                return HttpResponses.forbidden();
+            }
+            
+            if (StringUtils.isBlank(newTokenName)) {
```

Comment:
Specific API tokens are not accessible via URL, so we don't need to worry about validating the names any more than this, right?

Reply:
Exactly, the name is just a way to help user to remember which token is used for what. Nothing depends on the name except the display.